lib/voice/VoiceConnection.js

```diff
@@ -820,26 +915,37 @@ class VoiceConnection extends EventEmitter {
     }
 
     _sendAudioFrame(frame) {
-        this.sendNonce.writeUInt16BE(this.sequence, 2);
-        this.sendNonce.writeUInt32BE(this.timestamp, 4);
-
-        if(Sodium) {
-            const MACBYTES = Sodium.crypto_secretbox_MACBYTES;
-            const length = frame.length + MACBYTES;
-            this.sendBuffer.fill(0, 12, 12 + MACBYTES);
-            frame.copy(this.sendBuffer, 12 + MACBYTES);
-            Sodium.crypto_secretbox_easy(this.sendBuffer.subarray(12, 12 + length), this.sendBuffer.subarray(12 + MACBYTES, 12 + length), this.sendNonce, this.secret);
-            this.sendNonce.copy(this.sendBuffer, 0, 0, 12);
-            return this.sendUDPPacket(this.sendBuffer.subarray(0, 12 + length));
-        } else {
-            const BOXZEROBYTES = NaCl.lowlevel.crypto_secretbox_BOXZEROBYTES;
-            const ZEROBYTES = NaCl.lowlevel.crypto_secretbox_ZEROBYTES;
-            const length = frame.length + BOXZEROBYTES;
-            this.sendBuffer.fill(0, BOXZEROBYTES, BOXZEROBYTES + ZEROBYTES);
-            frame.copy(this.sendBuffer, BOXZEROBYTES + ZEROBYTES);
-            NaCl.lowlevel.crypto_secretbox(this.sendBuffer, this.sendBuffer.subarray(BOXZEROBYTES), ZEROBYTES + frame.length, this.sendNonce, this.secret);
-            this.sendNonce.copy(this.sendBuffer, BOXZEROBYTES - 12, 0, 12);
-            return this.sendUDPPacket(this.sendBuffer.subarray(BOXZEROBYTES - 12, BOXZEROBYTES + length));
+        const headerSize = this.sendHeader.length;
+        this.sendHeader.writeUInt16BE(this.sequence, 2);
+        this.sendHeader.writeUInt32BE(this.timestamp, 4);
+
+        this.#nonce = (this.#nonce + 1) >>> 0;
+        this.#nonceBuffer.writeUInt32BE(this.#nonce, 0);
+
+        if(this.mode === "aead_aes256_gcm_rtpsize") {
+            const cipher = crypto.createCipheriv("aes-256-gcm", this.secret, this.#nonceBuffer);
+            cipher.setAAD(this.sendHeader);
+            const result = Buffer.concat([cipher.update(frame), cipher.final(), cipher.getAuthTag()]);
+            this.sendHeader.copy(this.sendBuffer, 0, 0, headerSize);
+            result.copy(this.sendBuffer, headerSize);
+            this.#nonceBuffer.copy(this.sendBuffer, headerSize + result.length, 0, 4); // nonce padding
+            return this.sendUDPPacket(this.sendBuffer.subarray(0, headerSize + result.length + 4));
+        } else if(Sodium) {
+            const ABYTES = Sodium.crypto_aead_xchacha20poly1305_ietf_ABYTES;
+            const length = frame.length + ABYTES;
+            this.sendBuffer.fill(0, headerSize, headerSize + ABYTES);
+            frame.copy(this.sendBuffer, headerSize + ABYTES);
+            Sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(
+                this.sendBuffer.subarray(headerSize, headerSize + length),
+                this.sendBuffer.subarray(headerSize + ABYTES, headerSize + length),
```

Comment:
i think this might be off (although it is working fine), the `ABYTES` should come after the encrypted data, not before it I think